### ISSUE TITLE
KeyError: 'Head nodes supplied to generator contain IDs not found in graph'

### ISSUE THREAD

**Issue by ocesp98** [stellargraph/stellargraph#657]
I was trying to implement Stellargraph - Directed GraphSAGE, but then following error occurred: KeyError: 'Head nodes supplied to generator contain IDs not found in graph'. I don't know how to fix this.

I made the network by using the following code:
```python
G = nx.Graph()
G.add_nodes_from(list(sample.client.values))
G.add_nodes_from(list(sample.merchant.values))                   
G.add_nodes_from(list(sample.transaction.values))
edgelist1 = list((zip(sample.transaction.values, sample.client.values)))
edgelist2 = list((zip(sample.transaction.values, sample.merchant.values)))
G.add_edges_from(edgelist1)
G.add_edges_from(edgelist2)
```

Then I split the data in train_data and test_data. I then transform all features to numeric features. I used one-hot encoding to transform the categoric variable 'country'. Furthermore, I dropped some columns in the train and test data. 
Moreover, I defined a variable node_features that represents the features used as node attributes (minus the target variable). 

Then I transformed the graph into a Stellargraph : 
```python
stelgr = sg.StellarDiGraph(G, node_features= node_features)
batch_size = 50; in_samples = [5, 2]; out_samples = [5, 2]
generator = DirectedGraphSAGENodeGenerator(stelgr, batch_size, in_samples, out_samples)
train_gen = generator.flow(train_data, train_targets, shuffle=False)
```

Then this error occurs: 
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-49-825ab36d20da> in <module>
----> 1 train_gen = generator.flow(train_data, train_targets, shuffle=False)

~\Anaconda3\lib\site-packages\stellargraph\mapper\node_mappers.py in flow(self, node_ids, targets, shuffle)
   1107 
   1108         """
-> 1109         return NodeSequence(self, node_ids, targets, shuffle=shuffle)
   1110 
   1111     def flow_from_dataframe(self, node_targets, shuffle=False):

~\Anaconda3\lib\site-packages\stellargraph\mapper\node_mappers.py in __init__(self, generator, ids, targets, shuffle)
    105         if any(n not in generator.graph for n in ids):
    106             raise KeyError(
--> 107                 "Head nodes supplied to generator contain IDs not found in graph"
    108             )
    109 

KeyError: 'Head nodes supplied to generator contain IDs not found in graph'
```

**Comment by huonw** [comment]
Hi @ocesp98, thanks for trying stellargraph!

That error means that `train_data` contains some things that aren't known as IDs in the graph. To be sure of what the problem is, it would be very helpful to have more info, e.g. the code you used to create `train_data`, and even some information about the `sample` value. This will allow us to execute the failing code and so reproduce it on our machine.

In any case, we can start debugging now: the error message isn't great, but one way to debug this would be to manually reproduce what the validation code is doing:

```python
for id in train_data:
    if id not in generator.graph:
        print("Missing:", id)
```

Finally, thanks for filing an issue; no matter what the conclusion is, it made me notice that `DirectedGraphSAGENodeGenerator` is missing from our documentation! I'm fixing that in #658, and until that merges, you can see the documentation at https://external-builds.readthedocs.io/html/stellargraph/658/api.html#stellargraph.mapper.DirectedGraphSAGENodeGenerator .

**Comment by ocesp98** [comment]
The data consists of an unnamed column (integer representing the row number of the data line), client (client ID), merchant (merchant ID), country, amount, datetime, accepted, fraud. I took a random sample of the data (the dataset is too large to run on my own laptop). 
I dropped the unnamed column and I added a transaction column.
sample['transaction'] = list(zip(sample.category, sample.country, sample.amount, sample.datetime, sample.accepted))
I used following code to build the network. The purpose is to make a transaction node type containing the transaction information. The transaction node is connected to the client and merchant node. The goal is to predict the fraud value of the transaction node (node classification)
G = nx.Graph()
G.add_nodes_from(list(sample.client.values))
G.add_nodes_from(list(sample.merchant.values))               
G.add_nodes_from(list(sample.transaction.values))
edgelist1 = list((zip(sample.transaction.values, sample.client.values)))
edgelist2 = list((zip(sample.transaction.values, sample.merchant.values)))
G.add_edges_from(edgelist1)
G.add_edges_from(edgelist2)
print("number of nodes: ", G.number_of_nodes())
print(G)

train_data, test_data = model_selection.train_test_split(
    sample, train_size=0.8, test_size=0.2, stratify=sample['fraud']
)
target_encoding = feature_extraction.DictVectorizer(sparse=False)
train_targets = target_encoding.fit_transform(train_data[["fraud"]].to_dict('records'))
test_targets = target_encoding.transform(test_data[["fraud"]].to_dict('records'))
#one-hot encode country in trainset
train_data['country'] = pd.Categorical(train_data['country'])
dfDummies = pd.get_dummies(train_data['country'])
train_data = pd.concat([train_data, dfDummies], axis=1)
All features are then transformed into numeric features
In the train_data, I then drop the transaction, country and fraud column

feature_names = train_data.columns[3:]
node_features = train_data[feature_names]

stelgr = sg.StellarDiGraph(G, node_features= node_features)
print(stelgr.info())
The output is: StellarDiGraph: Directed multigraph
 Nodes: 22637, Edges: 40000

 Node types:
  default: [22637]
    Edge types: default-default->default

 Edge types:
    default-default->default: [40000]

batch_size = 50; in_samples = [5, 2]; out_samples = [5, 2]
generator = DirectedGraphSAGENodeGenerator(stelgr, batch_size, in_samples, out_samples)
generator.graph
for id in train_data:
    if id not in generator.graph:
        print("Missing:", id)
Missing: unnamed
Missing: client
Missing: merchant
Missing: category
Missing: amount
Missing: datetime
Missing: accepted
Missing: ARE
Missing: ARG
Missing: ARM
Missing: AUS
Missing: AUT
Missing: AZE
Missing: BEL
Missing: BGR
Missing: BRA
Missing: BWA
Missing: CAN
Missing: CHE
Missing: CHN
Missing: CYP
Missing: CZE
Missing: DEU
Missing: DNK
Missing: ESP
Missing: EST
Missing: FIN
Missing: FRA
Missing: GBR
Missing: GIB
Missing: GRC
Missing: HKG
Missing: HUN
Missing: IND
Missing: IRL
Missing: ISL
Missing: ISR
Missing: ITA
Missing: JPN
Missing: KNA
Missing: KOR
Missing: KWT
Missing: LUX
Missing: LVA
Missing: MAC
Missing: MAR
Missing: MEX
Missing: MLT
Missing: MUS
Missing: MYS
Missing: NGA
Missing: NLD
Missing: NOR
Missing: NZL
Missing: PAN
Missing: PER
Missing: PHL
Missing: POL
Missing: PRI
Missing: PRT
Missing: REU
Missing: ROU
Missing: RUS
Missing: SAU
Missing: SGP
Missing: SWE
Missing: THA
Missing: TUN
Missing: TUR
Missing: TWN
Missing: UKR
Missing: USA
Missing: VAT
Missing: VNM
Missing: ZAF
Missing: ZMB
So basically everything is missing from the StellarGraph

**Comment by huonw** [comment]
Ah, you're going to have do the investigation to confirm, but those values look like the column headings, rather than node IDs. The first argument to `flow` needs to be an iterable of node IDs, rather than all the data. I think this might be possible in your code using something like `train_data.transaction.values`:

```python
train_gen = generator.flow(train_data.transaction.values, train_targets, shuffle=False)
```

---

By the way, it's very difficult to read code in comments if it has no formatting; for example, the indentation is lost. Github's markdown offers a few ways to format code: https://help.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#quoting-code . I edited your comments to wrap the code in backticks so that I could help more easily, like:

~~~markdown
```python
... code here ...
```
~~~

**Comment by PantelisElinas** [comment]
Hi @ocesp98 in addition to what @huonw suggested, have a look at the following.

You create your nx graph using,

```
G.add_nodes_from(list(sample.client.values))
G.add_nodes_from(list(sample.merchant.values))               
G.add_nodes_from(list(sample.transaction.values))
```

The above create nodes with IDs given by the union of values in `sample.client.values`, `sample.merchant.values`, and `sample.transaction.values`. 

Your `node_features` DataFrame that you pass to `sg.StellarDiGraph(G, node_features= node_features)` should have an index that corresponds to the union of the node IDs in your graph. We use node IDs to lookup the corresponding feature vector in the DataFrame. If the set of node IDs and the DataFrame index disagree, then you will get the error you report. In other words, `list(G.nodes)` and `node_features.index` should be the same. Can you verify that this is the case?

For more information on how to create your `StellarGraph` object have a look at the documentation at,

https://stellargraph.readthedocs.io/en/stable/api.html#module-stellargraph.core

Regards,

P.

**Comment by ocesp98** [comment]
Then the problem is that the node features are not numeric
I have this 
![image](https://user-images.githubusercontent.com/59799466/72339738-1eb82d80-36c7-11ea-81ac-a9c8a88e1487.png)


I can make everything in the list numeric (then you will have a list of numeric values), but I'm not sure if StellarGraph will accept this type of input. Furthermore, the client and merchant IDs are of the object type and contain numbers as well as letters and sometimes even different signs like * or +

**Comment by PantelisElinas** [comment]
Hi @ocesp98 

just to clarify, node IDs don't have to be integers and in your case, for the client and merchant IDs, sounds like using strings should work.

With regards to the node features, you should prepare your data for StellarGraph the same way you would prepare your data for any non-graph ml algorithm and, yes, your node features should be numeric. If you have categorical features then you should one-hot encode them; you should also encode dates appropriately. 

Regards,

P.

**Comment by ocesp98** [comment]
Thanks for the help! I think it works. I only have the problem that I have more IDs than targets given that I artificially created transaction nodes inbetween the client and merchant nodes containing the transaction information. So, the goal is to predict whether the transaction node is fraudulent or not. Then my client and merchant nodes don't have any targets, but are still in the graph. ValueError: The length of the targets must be the same as the length of the ids

**Comment by huonw** [comment]
For supervised training like in this case, you need to pass in only the node IDs for which you have an annotated target. Nodes that don't have any ground truth data about whether they're fraudulent or not cannot be used for training, and so cannot be used in the call to `flow`.

**Comment by PantelisElinas** [comment]
Hi @ocesp98 

as @huonw suggested, I think you should check that in the before the following call,

```train_gen = generator.flow(train_data, train_targets, shuffle=False)```

you have `len(train_data) == len(train_targets)`. The `train_data` should be the node IDs and `train_targets` the true labels, one for each node ID.

Regards,

P.

**Comment by ocesp98** [comment]
Thank you very much for your help! It seems like the train generator now works

**Comment by ocesp98** [comment]
Okay, so I got a bit further in the code. Now I'm stuck with the model.fit_generator. I used the directed graphsage notebook as example for my code. 

````
graphsage_model = DirectedGraphSAGE(
    layer_sizes=[32, 32],
    generator = train_gen,
    bias=False,
    dropout=0.5,
)

x_inp, x_out = graphsage_model.build()
prediction = layers.Dense(units=train_targets.shape[1], activation="softmax")(x_out)

model = Model(inputs=x_inp, outputs=prediction)
model.compile(
    optimizer=optimizers.Adam(lr=0.005),
    loss=losses.categorical_crossentropy,
    metrics=["acc"],
)

history = model.fit_generator(
    train_gen,
    epochs=20,
    verbose=2,
    shuffle=False
)

Then I get this error: 

> ---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-63-0f1e25bf4aba> in <module>
      3     epochs=20,
      4     verbose=2,
----> 5     shuffle=True
      6 )

~\Anaconda3\lib\site-packages\tensorflow\python\keras\engine\training.py in fit_generator(self, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
   1431         shuffle=shuffle,
   1432         initial_epoch=initial_epoch,
-> 1433         steps_name='steps_per_epoch')
   1434 
   1435   def evaluate_generator(self,

~\Anaconda3\lib\site-packages\tensorflow\python\keras\engine\training_generator.py in model_iteration(model, data, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch, mode, batch_size, steps_name, **kwargs)
    218     step = 0
    219     while step < target_steps:
--> 220       batch_data = _get_next_batch(generator, mode)
    221       if batch_data is None:
    222         if is_dataset:

~\Anaconda3\lib\site-packages\tensorflow\python\keras\engine\training_generator.py in _get_next_batch(generator, mode)
    360   """Retrieves the next batch of input data."""
    361   try:
--> 362     generator_output = next(generator)
    363   except (StopIteration, errors.OutOfRangeError):
    364     return None

~\Anaconda3\lib\site-packages\tensorflow\python\keras\utils\data_utils.py in get(self)
    783     except Exception:  # pylint: disable=broad-except
    784       self.stop()
--> 785       six.reraise(*sys.exc_info())
    786 
    787 

~\Anaconda3\lib\site-packages\six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~\Anaconda3\lib\site-packages\tensorflow\python\keras\utils\data_utils.py in get(self)
    777     try:
    778       while self.is_running():
--> 779         inputs = self.queue.get(block=True).get()
    780         self.queue.task_done()
    781         if inputs is not None:

~\Anaconda3\lib\multiprocessing\pool.py in get(self, timeout)
    655             return self._value
    656         else:
--> 657             raise self._value
    658 
    659     def _set(self, i, obj):

~\Anaconda3\lib\multiprocessing\pool.py in worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
    119         job, i, func, args, kwds = task
    120         try:
--> 121             result = (True, func(*args, **kwds))
    122         except Exception as e:
    123             if wrap_exception and func is not _helper_reraises_exception:

~\Anaconda3\lib\site-packages\tensorflow\python\keras\utils\data_utils.py in get_index(uid, i)
    569       The value at index `i`.
    570   """
--> 571   return _SHARED_SEQUENCES[uid][i]
    572 
    573 

~\Anaconda3\lib\site-packages\stellargraph\mapper\node_mappers.py in __getitem__(self, batch_num)
    198         else:
    199             # Get sampled nodes for Attri2VecNodeGenerator
--> 200             batch_feats = self.generator.sample_features(head_ids)
    201 
    202         return batch_feats, batch_targets

TypeError: sample_features() missing 1 required positional argument: 'sampling_schema'

**Comment by huonw** [comment]
Unfortunately, I think that is a bug in stellargraph 0.8.3, there was a missing piece in some code that handled `DirectedGraphSAGENodeGenerator`, and so it fails. I've filed a separate issue #695 covering that specific issue, and opened a PR fixing it https://github.com/stellargraph/stellargraph/pull/696: we're planning to do a 0.8.4 release over the next day or two that will incorporate this fix.

Here's some temporary alternatives, that may or may not apply to your situation:

- one of the undirected algorithms such as `GraphSAGE`/`GraphSAGENodeGenerator` or `HinSAGE`/`HinSageNodeGenerator`
- install `stellargraph` from the branch in the PR above:
  - `pip install git+https://github.com/stellargraph/stellargraph.git@bugfix/695-directedgraphsage`
- install the latest development `stellargraph`:
  - You can install directly with `pip` (`pip install git+https://github.com/stellargraph/stellargraph.git`) or by cloning it manually: https://github.com/stellargraph/stellargraph#install-stellargraph-from-github-source
  - Note: this is a work-in-progress development version so other things may not work yet, and, there's been significant changes between 0.8.3 and the development version (https://github.com/stellargraph/stellargraph/blob/develop/CHANGELOG.md#head has a partial listing)

**Comment by huonw** [comment]
[0.8.4](https://github.com/stellargraph/stellargraph/releases/tag/v0.8.4) has now been released.

**Comment by huonw** [comment]
Hi @ocesp98, have we addressed your problem here?
